src/java/org/apache/cassandra/config/DataRateSpec.java

```diff
@@ -0,0 +1,376 @@
+        // doing the conversion in both directions.
+        return unit.convert(other.quantity, other.unit) == quantity && other.unit.convert(quantity, unit) == other.quantity;
+    }
+
+    @Override
+    public String toString()
+    {
+        return Math.round(quantity) + unit.symbol;
+    }
+
+    public enum DataRateUnit
+    {
+        BYTES_PER_SECOND("B/s")
+        {
+            public double toBytesPerSecond(double d)
+            {
+                return d;
+            }
+
+            public double toKibibytesPerSecond(double d)
+            {
+                return d / 1024;
+            }
+
+            public double toMebibytesPerSecond(double d)
+            {
+                return d / (1024 * 1024);
+            }
+
+            public double toMegabitsPerSecond(double d) { return d / 125000; }
+
+            public double convert(double source, DataRateUnit sourceUnit)
+            {
+                return sourceUnit.toBytesPerSecond(source);
+            }
+        },
+        KIBIBYTES_PER_SECOND("KiB/s")
+        {
+            public double toBytesPerSecond(double d)
+            {
+                return x(d, 1024, MAX / 1024.0);
+            }
+
+            public double toKibibytesPerSecond(double d)
+            {
+                return d;
+            }
+
+            public double toMebibytesPerSecond(double d)
+            {
+                return d / 1024;
+            }
+
+            public double toMegabitsPerSecond(double d)
+            {
+                return d / 122;
+            }
+
+            public double convert(double source, DataRateUnit sourceUnit)
+            {
+                return sourceUnit.toKibibytesPerSecond(source);
+            }
+        },
+        MEBIBYTES_PER_SECOND("MiB/s")
+        {
+            public double toBytesPerSecond(double d)
+            {
+                return x(d, 1024 * 1024, MAX / (1024.0 * 1024.0));
+            }
+
+            public double toKibibytesPerSecond(double d)
+            {
+                return x(d, 1024, MAX / (1024.0));
+            }
+
+            public double toMebibytesPerSecond(double d)
+            {
+                return d;
+            }
+
+            public double toMegabitsPerSecond(double d)
+            {
+                if (d > MAX / (MEGABITS_PER_MEBIBYTE))
+                    return MAX;
+                return Math.round(d * MEGABITS_PER_MEBIBYTE);
+            }
+
+            public double convert(double source, DataRateUnit sourceUnit)
+            {
+                return sourceUnit.toMebibytesPerSecond(source);
+            }
+        };
+
+        static final double MAX = Long.MAX_VALUE;
+        static final double MEGABITS_PER_MEBIBYTE = 8.388608;
+
+        /**
+         * Scale d by m, checking for overflow. This has a short name to make above code more readable.
+         */
+        static double x(double d, double m, double over)
+        {
```

Comment:
trivial: some way to add an `assert …` here to catch when over is not properly defined to prevent an overflow/POSITIVE_INFINITY ?

Reply:
over is used three times with specific constants and there is no expectation to variate so I don't think we really need it

Reply:
was just a suggestion. it's defensive programming, who knows what the next coder is thinking… (the method is open for use, and extension for further DataRateUnit classes is feasible…)

Reply:
Probably something like that should be reasonable to keep us in the right boundaries - `assert over < (MAX - 1) && (over == MAX / m);`

Reply:
we can probably also check for being positive, something like that

Reply:
thanks.

what came to mind as an example was CASSANDRA-3492

this was one of my first encounters with the Cassandra dev community.
a simple bug where units were not translated correctly lead to a downed production cluster (worse than that, every node has "corrupted" sstables and could not be started).

the consequences of an overflow or POSITIVE_INFINITY could be as severe.

Reply:
Agree, thank you for being patient and the example